### PR TITLE
[MER-41] Add ability to attach learning objectives to adaptive lessons

### PR DESCRIPTION
**Pull request by marc-hughes**
This adds the ability to attach learning objectives from the right hand "Screen" tab of adaptive lessons.

image

Clicking that button opens an editor to select them:

image

And you can see them from the scoring overview window:

image

Demo of how to use it:

https://user-images.githubusercontent.com/333265/178521121-c3e38e3d-0f6d-4e09-9b01-3b344963613f.mp4

### CODE REVIEW

File: assets/src/apps/authoring/components/PropertyEditor/custom/VariableEditor.tsx [hunk 1]
```diff
@@ -19,26 +19,22 @@ interface CustomFieldProps {
 }
 
 export const FieldTemplate: React.FC<any> = (props) => {
-  return (
-    <>
```
Comment (marc-hughes) [inline assets/src/apps/authoring/components/PropertyEditor/custom/VariableEditor.tsx:23]: Unnecessary fragment tag.

File: assets/src/apps/authoring/components/PropertyEditor/schemas/screen.ts [hunk 2]
```diff
@@ -135,6 +145,9 @@ export const screenUiSchema: UiSchema = {
       classNames: 'col-12',
     },
   },
+  learningObjectives: {
+    'ui:widget': LearningObjectivesEditor,
```
Comment (marc-hughes) [inline assets/src/apps/authoring/components/PropertyEditor/schemas/screen.ts:149]: the ui:widget on an array type is what needed the rjsf upgrade.

File: assets/src/apps/authoring/store/activities/actions/saveActivity.ts [hunk 3]
```diff
@@ -23,80 +23,86 @@ import { diff } from 'deep-object-diff';
 export const saveActivity = createAsyncThunk(
   `${ActivitiesSlice}/saveActivity`,
   async (payload: { activity: IActivity; undoable?: boolean }, { dispatch, getState }) => {
-    const { activity, undoable = true } = payload;
-    const rootState = getState() as any;
-    const projectSlug = selectProjectSlug(rootState);
-    const resourceId = selectResourceId(rootState);
-    const currentActivityState = selectActivityById(rootState, activity.id) as IActivity;
-    const group = selectCurrentGroup(rootState);
+    try {
```
Comment (marc-hughes) [inline assets/src/apps/authoring/store/activities/actions/saveActivity.ts:26]: Errors in here were getting swallowed with no indication there was an error, now at least a console message goes out.

File: assets/src/apps/delivery/store/features/groups/actions/deck.ts [hunk 4]
```diff
@@ -321,7 +321,7 @@ export const navigateToNextActivity = createAsyncThunk(
         return;
       }
     } else {
-      navError = `Current Activity ${currentActivityId} not found in sequence`;
+      navError = `deck uence`;
```
Comment (marc-hughes) [inline assets/src/apps/delivery/store/features/groups/actions/deck.ts:324]: todo - revert that

File: assets/src/apps/delivery/store/features/groups/actions/deck.ts [hunk 5]
```diff
@@ -56,7 +56,7 @@ export const initializeActivity = createAsyncThunk(
     const currentSequenceId = sequence.find((entry) => entry.activity_id === activityId)?.custom
       .sequenceId;
     if (!currentSequenceId) {
-      throw new Error(`Activity ${activityId} not found in sequence!`);
+      throw new Error(`deck::initializeActivity - Activity ${activityId} not found in sequence!`);
```
Comment (marc-hughes) [inline assets/src/apps/delivery/store/features/groups/actions/deck.ts:59]: I was trying to track down a `Activity ${activityId} not found in sequence` console log, but there were several of them, so I added a bit of context

File: assets/src/data/persistence/objectives.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ * But in adaptive lessons, we want to attach them to an activity, so
+ * we're going to "cheat" and attach them to the first part in the activity.
+ * However there's a few use cases we need to handle:
+ *
+ * 1. If there are no parts, a __default part is created, so attach it to that.
+ * 2. If the part that objectives are attached to is deleted, we should move the objectives
+ *    to another part.
+ * 3. If there were no parts, and objectives were attached, and later a part is added, that
+ *    default part is deleted, so move the objectives to the newly created part.
+ *
```
Comment (bsparks) [inline assets/src/data/persistence/objectives.ts:14]: just curious about the reasoning behind not using a dedicated `__objectives` part as opposed to managing the part shifts

Comment (marc-hughes) [reply assets/src/data/persistence/objectives.ts:14]: It didn't seem harder to do it this way, and results in fewer extra nodes in the data. I'm assuming for very close to 100% of actual real-life usage, we don't need __default or __objectives in a real scorable activity.

I also had some worry about analytics not picking up the part some day, but in the end, that was probably unfounded.

**Review by bsparks — APPROVED** [review]

**Pull request merged**